### ISSUE TITLE
Load/store R32/R64 types

### ISSUE THREAD

**Issue by yurydelendik** [CraneStation/cranelift#1176]
Most likely global.set/.get or table.set/.get for reference types will be translated into regular memory load and store instructions. Currently, their implementations are missing at codegen side. The https://github.com/CraneStation/cranelift/pull/1073 performs the translation of the wasm, but CL validator fails when ref types and globals/tables operations are present (see e.g. [example](https://github.com/WebAssembly/wasm-c-api/blob/master/example/hostref.wat))

**Comment by fitzgen** [comment]
The question is how to deal with barriers for e.g. generational gc or reference counting.

@sunfishcode and I had talked about how to handle this and one idea would be to have a trait method in `cranelift_wasm::FuncEnvironment` for translating table get/set to some snippet of IR that includes the barriers. This would mean that if one wasn't using `cranelift-wasm`, one would have to figure out their own method of introducing the barriers. This approach is flexible and maintains cranelift's current property of not calling back into the embedder when compiling IR to machine code.

Another variant of this is to introduce something akin to `FuncEnvironment` to `cranelift` proper, rather than only at the wasm translation layer. This would mean we are introducing embedder callbacks into compilation, and that seems like something that we would want to be careful about. But the conceptual advantage is that barriers are pretty low-level, and right now, after wasm translation the IR is very high level and then legalization passes bring it lower and lower level, so conceptually it would be congruent to legalize high-level reference type loads and stores into low-level loads and stores with barriers in the same phase.

The other idea we were oscillating back to was supporting just the three specific use cases we know we want to support: naive refcounting, spidermonkey gc, and a wasmtime gc. This avoids the full flexibility of the previous options and their bit of "architecture astronautics". I think this ran against the grain of some constraints that the .NET embedding has (+cc @peterhuene; or was it the Wasm C API?). I don't quite remember the details however. I vaguely remember coming away thinking that, yes, we would likely be better off with the fully general solution.

**Comment by bnjbvr** [comment]
> cranelift_wasm::FuncEnvironment for translating table get/set to some snippet of IR that includes the barriers.

For what it's worth, that's the way I envisioned it too; it would probably be sufficient for Spidermonkey.

**Comment by peterhuene** [comment]
With respect to embedding, some hosts will need to know when a reference is stored outside of their visibility (i.e. keeping an object on a GC heap reachable).  In the case of a .NET host, a [GCHandle](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle?view=netcore-3.0) will be allocated to keep the underlying GC object alive as it's passed as a Wasm reference type.  As this "handle" lives outside of the GC heap, it must be manually deallocated and, thus, reference counted to ensure the correct lifetime of the underlying GC object.

I was envisioning that the Rust and C embedding APIs having an option, probably at the "engine" level, to add additional IR for table.set/global.set that would invoke a callback set by the host.  Primarily, the host needs to know when a reference type is being stored, along with information about what is currently stored in the global / table slot.  I believe this would be sufficient for most hosts to provide generic reference counting of reference types.

I don't think loads will be important to the host, nor is tracking the reference type inside of Wasm frames, as the host will have opportunities to do the tracking at the boundaries between host and Wasm frames.

**Comment by fitzgen** [comment]
FYI, I've been evaluating how LLVM's GC support handles barriers as well as the interfaces JVMs use to support swapping out different GC implementations which have different barriers. Writing up a proposal based on that and I'll share it soon.

**Comment by fitzgen** [comment]
# GC Barriers and Cranelift





- [Why We Need To Support GC Barriers in Cranelift](#why-we-need-to-support-gc-barriers-in-cranelift)
- [Background](#background)
  - [Tri-Color Nomenclature](#tri-color-nomenclature)
  - [Example Barriers](#example-barriers)
- [Desiridata](#desiridata)
  - [Support Different GCs and GC Algorithms](#support-different-gcs-and-gc-algorithms)
  - [Support Fast Barrier Implementations](#support-fast-barrier-implementations)
  - [Elide and Coalesce Barriers (Eventually)](#elide-and-coalesce-barriers-eventually)
  - [Don't Lose Our Copy-and-Paste-This-IR-and-it-is-a-Test Property](#dont-lose-our-copy-and-paste-this-ir-and-it-is-a-test-property)
- [Prior Art](#prior-art)
  - [JEP 304: Garbage Collector Interface](#jep-304-garbage-collector-interface)
  - [Garbage Collection with LLVM](#garbage-collection-with-llvm)
  - [Discussion](#discussion)
- [Proposed Solution](#proposed-solution)
  - [The `store.ref` Instruction](#the-storeref-instruction)
    - [Immediates](#immediates)
    - [Parameters](#parameters)
    - [Semantics](#semantics)
  - [Inline IR Templates](#inline-ir-templates)
- [Open Questions](#open-questions)
- [Alternatives](#alternatives)
- [Implementation Plan](#implementation-plan)



## Why We Need To Support GC Barriers in Cranelift

Wasm `anyref`s can be managed by a GC (including reference counting) and Wasm will get full GC support in the future.

Cranelift can already precisely identify on-stack GC roots via stack maps. This means that we don't need to maintain the set of on-stack roots with a barrier the way SpiderMonkey's `Rooted<T>` does. So why do we need GC barrier support in Cranelift? A Wasm instance can store GC objects in places other than on the stack:

* Globals
* Tables

For example, if `anyref`s are reference counted in this embedding, and an `anyref` is assigned to a global, we need to decrement the reference count of the global's old value (if any) and increment the reference count of the globals new value:

```python
assign_anyref_to_global(global, new_value):
    increment reference count on new_value
    if global is not null:
        decrement reference count on *global
    *global = new_value
```

This is an example of a write barrier. Cranelift needs to enable its embedders to insert these snippets of code, or otherwise arrange for them to be executed. We would need similar write barrier support for if `anyref`s were managed by a generational and/or incremental collector (aka SpiderMonkey GC).

## Background

### Tri-Color Nomenclature

In the common tri-color nomenclature for GC, when tracing:

* **White**: An object that hasn't been seen yet.
* **Gray**: An object that has been seen, but we are not done processing its outgoing edges yet. It is effectively enqueued on the mark stack.
* **Black**: An object that we have finished processing.

### Example Barriers

For generational collectors, the heap is split into a tenured space for older objects and a nursery for newly allocated objects. The nursery is collected independently of the tenured space. This means that any edges from tenured objects to nursery objects must be treated as roots when collecting just the nursery. The "remembered set" of tenured->nursery edges are maintained (sometimes conservatively over-approximated) by a write barrier.

This is a typical post-write barrier for generational GC, similar to what is used by SpiderMonkey:

```python
generational_gc_write_barrier(receiver, slot, referent):
    receiver[slot] = referent
    if receiver is in tenured heap and referent is in nursery:
        call slow_path_add_to_remembered_set(receiver, slot, referent)
```

For concurrent collectors (not necessarily running the collector and mutator in parallel, but at least interleaving their execution) if a black object that the GC is already finished processing is ever mutated to point to a white object that the GC has not seen, and this becomes the only edge to the white object, then no other object will ever trace that white object.

The snapshot-at-the-beginning barrier, also known as a Yuasa barrier, is used for concurrent GCs. It is a pre-write barrier. SpiderMonkey uses this style of barrier for its incremental GC.

```python
snapshot_at_the_beginning_concurrent_gc_barrier(receiver, slot, referent):
    if receiver is gray and referent is white:
        call slow_path_mark_gray(referent)
    receiver[slot] = referent
```

The Steele and Dijkstra barriers are similar to each other. They are post-write barriers for concurrent GCs. The Steele barrier "undoes" work done by the GC by turning a black object gray. The Dijkstra barrier has the mutator do some extra work for the GC's tracing inside its barrier by enqueuing the referent on the mark stack.

```python
steele_concurrent_gc_barrier(receiver, slot, referent):
    receiver[slot] = referent
    if receiver is black and referent is white:
        call slow_path_mark_gray(receiver)

dijkstra_concurrent_gc_barrier(receiver, slot, referent):
    receiver[slot] = referent
    if receiver is black and referent is white:
        call slow_path_mark_gray(referent)
```

## Desiridata

### Support Different GCs and GC Algorithms

Some collectors / algorithms we know we want to support:

* SpiderMonkey GC (incremental and generational; has both pre-write and post-write barriers; if we include various weak caches, it also has read barriers, but these are used rarely, and are only used in C++ VM code AFAIK)

* Naive reference counting GC (with atomic reference counting and not) for simple Wasmtime embeddings with `anyref`s but without full Wasm GC

* An out-of-the-box tracing GC for Wasmtime with full Wasm GC types (maybe Boehm? not clear yet what this GC/algorithm will be and what barriers it will require!)

* Integrate with a host GC when embedding Wasmtime in a system that already has a GC, eg embedding Wasmtime in the Python interpreter.

Generally, we would like to be able to easily tweak barrier implementations, and experiment with different collectors and algorithms, similar to the way that you can swap out different GCs in many JVMs.

### Support Fast Barrier Implementations

Making barriers fast requires dividing fast, inline paths from slow, out-of-line paths and carefully balancing what operations fall on which side.

For generational GC, we might have the is-the-referent-in-the-nursery check inline, and only call the out-of-line slow path that implements the full barrier if that is the case:

```wat
;; receiver[slot] = referent
;; ... do stuff ...

ebb0(receiver, slot, referent):
    store referent to receiver[slot]
    v0 = referent is in nursery?
    brz v2, ebb1
    call out_of_line_slow_path(receiver, slot, referent)
    fallthrough ebb1

ebb1:
    ... do stuff ...
```

To support these inline checks, we need support for emitting custom barrier code around GC stores. This "triangle" control flow requires that the custom barrier-insertion code be able to split EBBs (or in a plain BB world, create BBs).

But barriers might be even more complicated: imagine a snapshot-at-the-beginning barrier for incremental GC that first has an inline check for whether GC is active (if not, we don't need to run a barrier at all) and if it is active, then jumps to a second layer stub that determines whether this store requires the full out-of-line call to remember this edge:

```wat
;; receiver[slot] = referent
;; ... do stuff ...

ebb0(receiver, slot, referent):
    v0 = is gc happening?
    brnz v0, stub
    fallthrough ebb1

ebb1:
    store referent to receiver[slot]
    ... do stuff ...

stub:
    v1 = is receiver gray?
    v2 = is referent white?
    v3 = v1 && v2
    brz v3, ebb1
    call out_of_line_slow_path(receiver, slot, referent)
    jmp ebb1
```

By pulling the stub out into its own block and having the we-don't-need-a-barrier case fall through, we avoid polluting the icache when GC is disabled. Creating this kind of barrier requires that the custom barrier-insertion code be able to split and create new blocks.

### Elide and Coalesce Barriers (Eventually)

There are various scenarios where barriers can be statically elided (often newly allocated objects don't need barriers) or where they can be coalesced (barriers are often idempotent, so we might not need to run a barrier N times for N mutations of the same object). Whether this is sound or not depends on the GC algorithm and barriers being used.

Initial support for GC barriers in Cranelift probably shouldn't implement these sorts of optimizations, but we want them eventually. We should be cognizant of that, and avoid designing ourselves into a corner where we can't support these optimizations in the future.

A few related resources:

* "Write Barrier Elision for Concurrent Garbage Collectors" by Vechev et al: https://researcher.watson.ibm.com/researcher/files/us-bacon/Vechev04Write.pdf

* "Compile-Time Concurrent Marking Write Barrier Removal" by Detlefs et al: https://www.cse.iitm.ac.in/~krishna/preprints/cgo05/cgo05.pdf

* https://webkit.org/blog/7122/introducing-riptide-webkits-retreating-wavefront-concurrent-garbage-collector/

### Don't Lose Our Copy-and-Paste-This-IR-and-it-is-a-Test Property

We have this nice property where we can always represent whatever IR we have in its text form, and therefore can dump IR at various points in between passes and use that dump as a test. If we aren't careful with how we design GC barrier integration, we might lose that property. For example, if GC barriers or their translation were 100% programmatic hooks (similar to `cranelift_wasm::FuncEnvironment`) and not represented at all in the IR.

It is worth pointing this out as an ability we would like to continue to have.

## Prior Art

There have been a lot of compilers for GC'd languages written, where the GC required cooperation from the compiler to insert barriers (or at least the inline, fast paths). What is a bit more rare is when a compiler intends to support multiple GCs/algorithms.

I am by no means aware of all the history here, and there are undoubtedly examples I'm missing. Nonetheless, these examples are from real world systems that seem to have been operating under similar constraints to us, and have had similar goals.

### JEP 304: Garbage Collector Interface

[JEP 304 introduced a interface that is implemented by Hotspot's GCs to disentangle them from the compiler's internals.][jep-304] That document describes the motivation and goals, but does not provide the implementation. Looking at the open-jdk sources, the most-relevant bits are [the `BarrierSet::AccessBarrier` class][access-barrier] and [the `BarrierSetAssembler` class][barrier-set-assembler]. The former is for VM code's barriers at runtime, the latter for emitting code for barriers in the JIT compiler. This parent class is used by the compiler when it is generating code to access an object and has methods for creating various kinds of loads and stores. The default parent class's implementations do not insert any barriers, just emit the code to perform the load or store operation. It is subclassed by the different GC implementations, who override a particular method to introduce a barrier before or after its load or store operation.

If we were to translate this scheme into Rust, it would be like having a `GcBarriers` trait that has methods for creating the code to do loads and stores and which the compiler calls at the appropriate places. This is similar to `cranelift_wasm::FuncEnvironment` and the way that it allows the embedder to provide custom Cranelift IR translations of calls, global accesses, etc. The `BarrierSetAssembler` class, used by the compiler, is full of virtual calls, which corresponds to a Rust trait object. The `BarrierSet::AccessBarrier` class, which is used at runtime by the VM, uses a bunch of template methods, which corresponds to generic functions with type parameters in Rust.

What I don't see here is any way for a GC implementation to describe to the compiler the conditions under which a barrier is redundant, and it's therefore safe to elide it. I'm not 100% sure, but I don't think Hotspot's compiler will elide redundant barriers.

[jep-304]: https://openjdk.java.net/jeps/304
[access-barrier]: https://github.com/unofficial-openjdk/openjdk/blob/7a481e8febeb8c497a60e62f230537becd0e28a9/src/hotspot/share/gc/shared/barrierSet.hpp#L173-L318
[barrier-set-assembler]: https://github.com/unofficial-openjdk/openjdk/blob/7a481e8febeb8c497a60e62f230537becd0e28a9/src/hotspot/cpu/x86/gc/shared/barrierSetAssembler_x86.hpp#L34-L89

### Garbage Collection with LLVM

[LLVM also has some support for integrating with GCs.][llvm-gc] LLVM does not itself provide a GC, just the hooks necessary for your LLVM-using compiler to integrate with your own, external GC. LLVM has seemingly put a lot of thought into how to represent GC safepoints and stack maps, and there is much less discussion of barriers.

LLVM represents loads and stores from the GC heap with dedicated intrinsic instructions that are distinct from regular memory loads and stores: `@llvm.gcread` and `@llvm.gcwrite`. By default, these instructions are eventually lowered to volatile loads and stores. However, LLVM embedders may provide a custom implementation of these intrinsics by writing and registering a custom lowering pass.

LLVM's documentation and code don't seem to have anything explicitly about eliding redundant barriers, although it could always be done manually via the embedder writing and registering a custom function pass (or integrating it with their existing barrier-lowering pass).

[llvm-gc]: https://releases.llvm.org/9.0.0/docs/GarbageCollection.html
[gcstrategy]: http://llvm.org/doxygen/classllvm_1_1GCStrategy.html

### Discussion

Both LLVM and JEP 304 allow for custom GC strategies with different barriers. Comparing and contrasting them shows a few inter-related questions in the design of supporting custom GC barriers:

* When are barriers lowered?
* What are barriers lowered into?
* How are barriers represented before they are lowered?

JEP 304 chose one end of the design space, where barriers are lowered as machine code is emitted. The barriers are lowered directly to machine code. The barriers are not explicitly represented in the IR, and implicitly wrap heap loads and stores.

LLVM chooses a different point in the design space where barriers are lowered in the middle of the compilation pipeline, and they are lowered into LLVM IR, which can then be subsequently processed by all of LLVM's optimization passes. Before lowering, the barriers are explicitly represented in the IR as intrinsic instructions.

An implied design that is missing from the prior art we've considered is to lower barriers at the beginning of the compilation pipeline, perhaps as the source language is translated into IR. The barriers are only represented as their raw implementation code that is inline. This would be equivalent to if `cranelift-wasm` translated a Wasm `global.set` of an `anyref` global into clif that included the raw code for the associated write barrier. This has the advantage of not requiring any cooperation from the code generator to emit barriers, but the embedder is completely on their own, and doesn't get any guiding hand from the code generator.

Making barriers explicit in the IR gives the potential for reasoning about them directly, for example writing a barrier elision pass. However, in a GC implementation-agnostic situation, this requires some cooperation between the code generator and embedder in the form of support for custom passes or some kind of hooks.

Lowering barriers into the code generator's IR allows both portability and subsequent optimization passes to operate over the barrier's expanded code, while lowering them to machine code directly gives embedders precise control over the emitted code.

## Proposed Solution

I propose that we:

* Introduce a `store.ref` instruction to explicitly represent potentially barrier-protected stores in the IR.

* Generalize the global value sub-language to arbitrary inline IR templates.

* Use these templates to lower `store.ref` instructions into IR that implements their raw load/store and barrier code.

This proposed solution is at a similar point in the design space as LLVM's GC support, but does not require support for custom passes to lower the barriered operations into IR. The advantage of generalizing the global value sub-language is that we don't have to introduce support for custom passes just to get GC support up and running, and we don't lose our current "copy and paste this IR and it's a test" abilities.

We would not initially support eliding barriers directly, although we would be able to run our few optimization passes and any future peephole passes over their expanded code. We do have a clear path forward for how to support this feature when the time comes: allow embedders to write their own custom passes.

This support for write barriers could be easily extended to support load barriers in the future as well. Load barrier support isn't included in this proposal because it simply isn't necessary for our initial goals.

### The `store.ref` Instruction

```
store.ref [memflags] value, receiver [+offset]
```

#### Immediates

* `memflags: (notrap|aligned|readonly)*` &mdash; optional memory flags for the store; same as regular memory stores.
* `offset: i32` &mdash; an optional 32 bit offset into the referenced GC object. Defaults to zero. Initially must be zero, but we will loosen that once we support full Wasm GC objects with fields.

#### Parameters

* `value: r32 | r64` &mdash; the new value being stored into `receiver + offset`.
* `receiver: r32 | r64` &mdash; the receiver object reference.

#### Semantics

```
*(receiver + offset) = value
```

This has the same semantics as if the receiver reference type was a regular address and this were a regular store, but it might expand into more code than a regular store would if the embedder's GC requires write barriers (eg it uses generational or concurrent GC).

### Inline IR Templates

I think @sunfishcode has his own (probably more fully fledged) ideas of how the global values sub-language could be generalized into always-inlined templates of arbitrary IR snippets. I'll just provide a sketch of one possible way we could do this.

Templates are *always* inlined, and as such, they must not be recursive. We can conservatively enforce this by only allowing the *n<sup>th</sup>* template to use templates *0* through *n-1*.

Here is a strawperson clif example that uses and inline IR template to implement a generational GC write barrier:

```
function %store_in_anyref_table(r32, i32) {
    table0 = some anyref table

    fn0 = %out_of_line_slow_path(r32, r32)

    ;; Takes the value, receiver, and offset and implements the write barrier.
    ;;
    ;; Imagine the nursery is known to occupy the address range 0x1000 through
    ;; 0x2000.
    ;;
    ;; Do an inline check if the new value is in the nursery, and if so, call
    ;; the slow out-of-line path for the barrier.
    templ0 = (r32, r32, i32) {
      ebb0(v0: r32, v1: r32, v2: i32):
        ;; Do the actual store.
        store v0, v1+v2

        ;; Check if the new value is in the nursery.
        v3 = bitcast.i32 v0
        v4 = icmp_imm uge v3, 0x1000
        v5 = icmp_imm ult v3, 0x2000
        v6 = band v4, v5

        ;; If so, go to our slow path.
        brnz ebb1

        ;; If not, continue on our merry way!
        return

      ebb1:
        ;; Slow path. We know the new value is in the nursery, so we may need to
        ;; record it in the remembered set. Handle this with an out of line call.
        call fn0(v0, v1)
        return
    }

    ;; Tell Cranelift that `store.ref` instructions should be lowered with
    ;; `templ0`.
    store.ref = templ0

ebb0(v0: r32, v1: i32):
    ;; Get an `r32` pointing to the table's `v1`th element.
    v2 = table_ref.r32 table0, v1

    ;; Store `v0` into the table. This will need a write barrier!
    store.ref v0, v2+0

    return
}
```

## Open Questions

* Taking and returning immediates and memory flags is maybe going to be a little funky in templates, but it is something we kind of want, since a `store.ref` could be annotated with memory flags and could eventually have a constant immediate offset. AFAIK, we don't really have any precedent in Cranelift for const function parameters, but this is sort of what we want.

* As an embedder, when actually implementing the barrier code that `store.ref`s should expand to, there will likely be memory loads and maybe also stores in the expanded code. These could be used in the process of checking whether an object is white/gray/black or if it is in the nursery or not. Right now we would have to use the equivalent of volatile loads and stores, so that if a GC happens our barriers don't operate on stale results. It would be nice to eventually have a kind of load and store where we *could* GVN away redundant loads, but where we would *not* do it across GC safe points.

* Barriers are currently the biggest missing piece in Cranelift needed to support Wasm `anyref`s, but just implementing support for them is not sufficient. We also need a way to get an `r32` or `r64` out of a global, which we don't currently support but shouldn't be too hard. We also need the equivalent of `table_addr` but for getting an `r32` or `r64` instead of an address for `anyref` tables. Or maybe we just override `table_addr` and make it polymorphic. Not clear to me exactly what is the best approach here, but it also seems like an easier problem than general GC barrier support, so I haven't thought too hard about it yet.

* An embedder might want slightly different write barriers for references that it knows are different kinds of objects, or for writing to a Wasm global vs writing to a Wasm `anyref` table. Should we attach the inline template that implements a write barrier to the `store.ref` instruction as an immediate, allowing different barriers for different stores?

## Alternatives

There is some complexity to the design of inline IR templates. There are two ways we could avoid IR templates and their design complexity:

1. By following LLVM's design even more closely and lowering `store.ref` into regular stores by default, but also providing a way for embedders to register custom passes that do a custom lowering of `store.ref` that includes their custom barriers.

2. By creating a trait similar to `cranelift_wasm::FuncTranslator` inside `cranelift_codegen` that the embedder implements, gives us, and we use to lower any `store.ref`s we encounter:

   ```rust
   pub trait GcBarriers {
       // Lower the `store.ref` at `pos` into a raw store and its associated
       // barrier.
       fn store_ref(&mut self, pos: &mut FuncCursor);
   }
   ```

   This is a bit more similar to JEP 304, but we'd still have an explicit representation of barriered operations in the IR and could still run peephole passes or whatever on the barrier implementation after lowering.

Both of these alternatives would cause us to lose the easy "copy and paste this IR as a test" ability we have now.

Option (1) happens to empower embedders to elide redundant GC barriers themselves, for what it's worth. I think it is a viable alternative, since allowing embedders to implement custom passes is the only practical way to elide redundant barriers; trying to encode all the necessary conditions into a GC implementation-agnostic generic trait interface won't work out.

## Implementation Plan

Assuming we all reach consensus, here is my breakdown of implementation steps:

1. Generalize the global value sub-language into inline IR templates.

2. Add the `store.ref` instruction and use inline IR templates to lower them.

3. Add `anyref` in tables support to `cranelift-wasm`, creating `store.ref` instructions.

4. In Wasmtime, get `anyref`s with interface types and reference counting working, kicking the tires on custom barriers in the process.

5. Now that the support for custom GC barriers has been proven out a little, someone else integrates with SpiderMonkey GC.

This can, of course, change depending on where we go with the design, or just based on any feedback in general.

**Comment by bnjbvr** [comment]
Great post, thanks! Tagging @eqrion who's been lately working on ref types integration in Spidermonkey, to make sure you're in sync with your approaches here. Overall a great plan, adding templates as a way to avoid codegen traits sounds good to me, and could be reused in other contexts. Most open questions below thus relate to templates themselves.

A few thoughts:

- Since WeakRefs are making their way into JS, I'd expect that we'd have them in wasm too, and they'll probably require the read barrier (later).

- Another possible idea for a (far away in the future) optimization would be to group barriers sequentially. This is slightly different from coalescing, in the sense that a codegen'd program may be, if the embedder decides it, stopped only at precise points, in addition to any place where a GC could happen (import calls, allocations...). In Spidermonkey this happens at function entries and loop headers, and any function call, if I'm not mistaken. Thus all the addresses would be buffered in, and the costly out-of-line call happens only once just before giving back control to the embedder (which then can safely run another GC slice). I don't know how worthwhile is this idea (probably only worthwhile if the buffer doesn't need explicit function calls to be manipulated, so its inline storage is a bit large?), or if this has been explored in the past.

- > It would be nice to eventually have a kind of load and store where we could GVN away redundant loads, but where we would not do it across GC safe points.

  Alias analysis is the way to go to allow distinguishing different kinds of loads/stores and apply different optimizations based on those. There would be an alias kind for GC objects, a different alias kind for others, and a safepoint would be a write of the GC object alias kind. If the alias kinds are controlled by the embedder, then it would also address the last item (`An embedder might want slightly different write barriers for references that it knows are different kinds of objects`).

- A small nit in naming, because who doesn't want a new bikeshed for end-of-year celebrations: the dot in instructions relates to type bindings, in the parser, so `store.ref` would be a bit ambiguous, `store_ref` would be a possible non-ambiguous replacement.

- While I'm at it, it'd be nice to have an extra text detail in the name of instructions bound to templates, to see immediately that they're different from regular instructions, when they come in the instruction flow: adding a one-letter prefix, for instance.

- Just to make sure it's mentioned somewhere: we want to also make sure that templates are not recursive through the binding of templates to instructions that might have been used in a template. A (stupid, but possible) example would be that the template used for `store_ref` make use of `store_ref` itself. (Templates could make use of instructions that are also bound to other templates, so it gets a bit more hairy than that.)

- One little thing worrying about templates: they need to exist once per function, if they were to be defined within a Cranelift function. For what it's worth, some wasm modules may have thousands of functions, so the memory overhead might be non-negligible, even if we're using a lazy approach to generate templates (when they're to be used). I understand this came from the requirement to maintain the "copy/paste" IR property; I would have imagined that using global (as in: outside the scope of functions) templates could exist too, and be added globally (as we do for settings, for instance), thus making templates a module-scoped concept. It might be the first one? But it does seem to make sense and would address the possible memory overhead while keeping the dump-text property. Then we can imagine that when setting the safepoints / reftypes in Cranelift, we must provide templates for each of `load_ref`, `store_ref`, even if they're empty, so the embedder doesn't shoot themselves in the feet: `set enable_safepoints true, $load_ref=%tmpl0, $store_ref=%tmpl1`.

**Comment by bjorn3** [comment]
> I would have imagined that using global (as in: outside the scope of functions) templates could exist too, and be added globally (as we do for settings, for instance), thus making templates a module-scoped concept.

`Rc<Template>` would avoid duplication of templates.

**Comment by eqrion** [comment]
As Benjamin mentioned, I've been implementing reftypes integration in SpiderMonkey/CL. I think you've got a good summary of the issues that exist in a full-featured implementation.

My work so far has been to just get the feature working, even if it's sub-optimal. For example, using VM functions for all table operations that include their own barriers, and using pre-barrier and post-barrier VM functions for global writes. This has worked well and has allowed me to focus on the long tail of issues needed to pass all our tests. Significantly, the stackmap representation is much different than SM's.

I did experiment with trying to emit a more optimal pre-barrier/post-barrier for global writes by splitting the fast-inline and slow-fallback parts. My approach involved tweaking `wasm::FuncEnvironment` to be able to emit local control flow constructs, and adding a special hook for `global.set`. I got it mostly working, but it was not going to be worth the complexity at that point. So currently, I just have a hook in `FuncEnvironment` for emitting the pre/post-barrier VM calls for `global.set`.

I don't think your proposed implementation plan will conflict with my work. I'll be happy to use whatever solution is chosen for emitting optimal barriers. If there isn't a solution in-tree by the time my work is ready to land, I believe the simple approach is low-complexity enough to land and port over as needed.

**Comment by hotsphink** [comment]
> SpiderMonkey GC (incremental and generational; has both pre-write and post-write barriers; if we include various weak caches, it also has read barriers, but these are used rarely, and are only used in C++ VM code AFAIK)

This isn't entirely true. SpiderMonkey very commonly uses read barriers in the interface with the Gecko embedding, for integrating with the cycle collector. Any embedding object that might be a  member of a collectible cycle will have a read barrier that prevents it from being cycle collected.

Specifically, `JS::Heap<T>` has a read barrier that does `UnmarkGrayGCThingRecursively` aka `ExposeGCThingToActiveJS`, though note that "gray" in that name is "CC gray", not at all the same as the "gray" in https://github.com/bytecodealliance/cranelift/issues/1176#issuecomment-566790342. It's not a weak reference, it's more a workaround for not having write barriers throughout the embedding.

I don't know if that's at all relevant here, though.

**Comment by sunfishcode** [comment]
In support of the plan here, I've now submitted https://github.com/bytecodealliance/cranelift/pull/1299 to start generalizing global values into templates.

**Comment by fitzgen** [comment]
First off: thanks for the detailed responses everyone!

-----------------------------------------------

@eqrion great! Sounds like we can work in parallel :)

A good incremental stepping stone might be using templates (once we have them working with `store_ref`) to simply emit a call to the out of line functions you're implementing now.

--------------------------------------------------

@bnjbvr

> Since WeakRefs are making their way into JS, I'd expect that we'd have them in wasm too, and they'll probably require the read barrier (later).

@hotsphink

> SpiderMonkey very commonly uses read barriers in the interface with the Gecko embedding, for integrating with the cycle collector. Any embedding object that might be a member of a collectible cycle will have a read barrier that prevents it from being cycle collected.

Great points! It probably does make sense to support read barriers from the get-go then.

Out of curiosity, does ion ever emit inline read barriers for objects in its JIT'd code, similar to those used by `JS::Heap<T>`, or do those read barriers only exist in C++ code?

--------------------------------------------------------

@bnjbvr

> There would be an alias kind for GC objects, a different alias kind for others, and a safepoint would be a write of the GC object alias kind. If the alias kinds are controlled by the embedder, then it would also address the last item (An embedder might want slightly different write barriers for references that it knows are different kinds of objects).

Yes, this would work perfectly.

> `store_ref`

Sure :)

**Comment by hotsphink** [comment]
> Out of curiosity, does ion ever emit inline read barriers for objects in its JIT'd code, similar to those used by `JS::Heap<T>`, or do those read barriers only exist in C++ code?

From my shaky understanding, there are no read barriers in JIT code. I think a plain property read, for example, would not need a read barrier, at least not until the value read makes its way out to the calling C++ Gecko code. SpiderMonkey code won't read barrier "internal" things even in C++ code, it relies on Gecko to read barrier things that get returned from the JS engine.

The JIT does generate code for both pre- and post-write barriers, when it can't prove they aren't needed.
